test/common/http/http2/codec_impl_fuzz_test.cc

```diff
@@ -0,0 +1,353 @@
+        headers.setReferenceKey(Headers::get().Status, "100");
+        encoder.encode100ContinueHeaders(headers);
+      }
+      break;
+    }
+    case test::common::http::http2::DirectionalAction::kHeaders: {
+      if (state == StreamState::PendingHeaders) {
+        encoder.encodeHeaders(fromHeaders(directional_action.headers()), end_stream);
+        state = end_stream ? StreamState::Closed : StreamState::PendingDataOrTrailers;
+      }
+      break;
+    }
+    case test::common::http::http2::DirectionalAction::kData: {
+      if (state == StreamState::PendingDataOrTrailers) {
+        Buffer::OwnedImpl buf(std::string(directional_action.data() % (1024 * 1024), 'a'));
+        encoder.encodeData(buf, end_stream);
+        state = end_stream ? StreamState::Closed : StreamState::PendingDataOrTrailers;
+      }
+      break;
+    }
+    case test::common::http::http2::DirectionalAction::kTrailers: {
+      if (state == StreamState::PendingDataOrTrailers) {
+        encoder.encodeTrailers(fromHeaders(directional_action.trailers()));
+        state = StreamState::Closed;
+      }
+      break;
+    }
+    case test::common::http::http2::DirectionalAction::kReset: {
+      if (state != StreamState::Closed) {
+        encoder.getStream().resetStream(
+            static_cast<Http::StreamResetReason>(directional_action.reset()));
+        request_state_ = response_state_ = StreamState::Closed;
+      }
+      break;
+    }
+    case test::common::http::http2::DirectionalAction::kReadDisable: {
+      if (state != StreamState::Closed) {
+        const bool disable = directional_action.read_disable();
+        if (read_disable_count_ == 0 && !disable) {
+          return;
+        }
+        if (disable) {
+          ++read_disable_count_;
+        } else {
+          --read_disable_count_;
+        }
+        encoder.getStream().readDisable(disable);
+      }
+      break;
+    }
+    default:
+      // Maybe nothing is set?
+      break;
+    }
+  }
+
+  void streamAction(const test::common::http::http2::StreamAction& stream_action) {
+    switch (stream_action.stream_action_selector_case()) {
+    case test::common::http::http2::StreamAction::kRequest: {
+      directionalAction(request_state_, *request_encoder_, stream_action.request());
+      break;
+    }
+    case test::common::http::http2::StreamAction::kResponse: {
+      directionalAction(response_state_, *response_encoder_, stream_action.response());
+      break;
+    }
+    default:
+      // Maybe nothing is set?
+      break;
+    }
+  }
+
+  NiceMock<MockStreamCallbacks> server_stream_callbacks_;
+  NiceMock<MockStreamDecoder> response_decoder_;
+  StreamEncoder* request_encoder_;
+  StreamEncoder* response_encoder_;
+  NiceMock<MockStreamDecoder> request_decoder_;
+  StreamState request_state_;
+  StreamState response_state_;
+  uint32_t read_disable_count_{};
+};
+
+// Buffer between client and server H2 codecs. This models each write operation
```

Comment:
I suppose at some point we might want to actually be able to reorder full h2 frames? I wonder if that would make the fuzzing more effective or not. I'm not sure.

Reply:
Yeah; based on experience, some of the time you do get full frame reordering due to 1 write per frame, but some of the time it's only partial. In theory, the fuzzer could generate a series of swap actions that performs a full frame swap, but the chance of that happening is lower as the frame fragmentation grows. I'm planning on looking at coverage and other metrics to try and tune this once we land in oss-fuzz and internal runs.